pkg/ipamd/rpc_handler.go

```diff
@@ -80,21 +75,12 @@ func (s *server) AddNetwork(ctx context.Context, in *rpc.AddNetworkRequest) (*rp
 }
 
 func (s *server) DelNetwork(ctx context.Context, in *rpc.DelNetworkRequest) (*rpc.DelNetworkReply, error) {
-	log.Infof("Received DelNetwork for Pod %s, Namespace %s, Sandbox %s",
-		in.K8S_POD_NAME, in.K8S_POD_NAMESPACE, in.K8S_POD_INFRA_CONTAINER_ID)
+	log.Infof("Received DelNetwork for Sandbox %s",
+		in.K8S_POD_INFRA_CONTAINER_ID)
 	delIPCnt.With(prometheus.Labels{"reason": in.Reason}).Inc()
 
-	ip, deviceNumber, err := s.ipamContext.dataStore.UnassignPodIPv4Address(&k8sapi.K8SPodInfo{
-		Name:      in.K8S_POD_NAME,
-		Namespace: in.K8S_POD_NAMESPACE,
-		Sandbox:   in.K8S_POD_INFRA_CONTAINER_ID})
+	ip, deviceNumber, err := s.ipamContext.dataStore.UnassignPodIPv4Address(in.K8S_POD_INFRA_CONTAINER_ID)
 
-	if err != nil && err == datastore.ErrUnknownPod {
-		// If L-IPAMD restarts, the pod's IP address are assigned by only pod's name and namespace due to kubelet's introspection.
-		ip, deviceNumber, err = s.ipamContext.dataStore.UnassignPodIPv4Address(&k8sapi.K8SPodInfo{
-			Name:      in.K8S_POD_NAME,
-			Namespace: in.K8S_POD_NAMESPACE})
-	}
```

Comment:
I always felt that this was super hacky. It was in my long backlog to clean it up, so thanks for working on this.

Can it ever be that `k8sArgs.K8S_POD_INFRA_CONTAINER_ID` is not set? I was referring to that old comment:

     If L-IPAMD restarts, the pod's IP address are assigned by only pod's name and namespace due to kubelet's introspection.

Reply:
I think that comment refers to the way the aws-node internals work.  In the current version of this code, it calls this function _without_ container ID to indicate ipam cold-restart discovery and _with_ container ID to indicate regular CNI-driven pod creation.

Fwiw Kubernetes' dockershim [*always*](https://github.com/kubernetes/kubernetes/blob/6acceaeb2dbfc4e4b4528eeee0b034de999213a6/pkg/kubelet/dockershim/network/cni/cni.go#L395-L406) sets K8S_POD_INFRA_CONTAINER_ID to the same as CNI (non-k8s) ContainerID arg.  I haven't audited other CRI implementations (I should do so).  I think we can rely on always using CNI ContainerID arg (that's the point), so I might switch to that.